frontend/packages/console-shared/src/hooks/useUserSettings.ts

```diff
@@ -0,0 +1,146 @@
+
+// This won't be needed once we have proxy api
+const getProject = async () => {
+  try {
+    await k8sGet(ProjectModel, USER_SETTING_CONFIGMAP_NAMESPACE);
+  } catch {
+    await k8sCreate(ProjectRequestModel, {
+      metadata: {
+        name: USER_SETTING_CONFIGMAP_NAMESPACE,
+      },
+    });
+  }
+};
+
+const createConfigMap = async (configMapData: K8sResourceKind): Promise<boolean> => {
+  try {
+    await k8sCreate(ConfigMapModel, configMapData);
+    return true;
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+    return false;
+  }
+};
+
+const updateConfigMap = async (configMap: ConfigMapKind, key: string, value: string) => {
+  if (value !== configMap.data?.[key]) {
+    const patch = [
+      {
+        op: 'replace',
+        path: `/data/${key}`,
+        value,
+      },
+    ];
+    try {
+      await k8sPatch(ConfigMapModel, configMap, patch);
+    } catch (err) {
+      // eslint-disable-next-line no-console
+      console.error(err);
+    }
+  }
+};
+
+const deseralizeData = (data: string) => {
+  try {
+    return JSON.parse(data);
+  } catch {
+    return data;
+  }
+};
+
+const seralizeData = <T>(data: T) => {
+  if (typeof data === 'string') {
+    return data;
+  }
+  return JSON.stringify(data);
+};
+
+export const useUserSettings = <T>(
+  key: string,
+  defaultValue?: T,
+): [T, React.Dispatch<React.SetStateAction<T>>, boolean] => {
+  const defaultValueRef = React.useRef<T>(defaultValue);
+  const keyRef = React.useRef<string>(key);
+  const userUid = useSelector(
+    (state: RootState) => state.UI.get('user')?.metadata?.uid ?? 'kubeadmin',
+  );
+  const configMapResource = React.useMemo(
+    () => ({
+      kind: ConfigMapModel.kind,
+      namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+      isList: false,
+      name: `user-settings-${userUid}`,
+    }),
+    [userUid],
+  );
+  const [cfData, cfLoaded, cfLoadError] = useK8sWatchResource<K8sResourceKind>(configMapResource);
+  const [settings, setSettings] = React.useState<T>();
+  const [loaded, setLoaded] = React.useState(false);
+
+  const configMapData = React.useMemo(
+    () => ({
+      apiVersion: ConfigMapModel.apiVersion,
+      kind: ConfigMapModel.kind,
+      metadata: {
+        name: `user-settings-${userUid}`,
+        namespace: USER_SETTING_CONFIGMAP_NAMESPACE,
+      },
+      data: {
+        [keyRef.current]: seralizeData(defaultValueRef.current),
+      },
+    }),
+    [userUid],
+  );
+
+  React.useEffect(() => {
+    if (cfLoadError || (!cfData && cfLoaded)) {
+      (async () => {
+        await getProject();
+        const cmCreated = await createConfigMap(configMapData);
+        if (!cmCreated) {
+          setSettings(defaultValueRef.current);
+          setLoaded(true);
+        }
+      })();
+    } else if (
+      cfData &&
+      cfLoaded &&
+      (!cfData.data?.hasOwnProperty(keyRef.current) ||
+        seralizeData(settings) !== cfData.data?.[keyRef.current])
```

Comment:
You still need to call `setLoaded(true)` even if these conditions fail and `cfLoaded === true`